developer-hub/src/reference_docs/defensive_programming.rs

```diff
@@ -0,0 +1,568 @@
+//! floating point calculations.
+//!
+//! Intentional and predictable design should be our first and foremost
+//! property for ensuring a well running, safely designed system.
+//!
+//! ## Defensive Programming
+//!
+//! Defensive programming is a design paradigm that enables a particular program to continue
+//! running despite unexpected behavior. These unforseen circumstances may
+//! cause the program to stop, (or in the Rust context, `panic!`) defensive practices allow for
+//! these circumstances to be accounted for ahead of time, and for them to be handled in a more
+//! graceful manner.
+//!
+//! The Polkadot SDK is both built to reflect these principles, and to be used accordingly.
+//!
+//! ## General Practices
+//!
+//! When developing within the context of the a runtime, there is one golden rule:
+//!
+//! ***DO NOT PANIC!***
+//!
+//! Most of the time - there are some exceptions, such as critical operations being actually more
+//! dangerous than allowing the node to continue running (block authoring, consensus, etc).
+//!
+//! - Directly using `unwrap()` for a [`Result`] shouldn't be used.
+//! - This includes accessing indices of some collection type, which may implicitly `panic!` (i.e.,
+//!   via `get()`)
+//! - It may be acceptable to use `except()`, but only if one is completely certain (and has
+//!   performed a check beforehand) that a value won't panic upon unwrapping.
+//! - If you are writing a function that could panic, [be sure to document it!](https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html#documenting-components)
+//! - Many seemingly, simplistic operations, such as **arithmetic** in the runtime, could present a
+//!   number of issues [(see more later in this document)](#integer-overflow).
+//!
+//! ### Defensive Traits
+//!
+//! To also aid in debugging and mitigating the above issues, there is a
+//! [`Defensive`](frame::traits::Defensive) trait (and its companions,
+//! [`DefensiveOption`](frame::traits::DefensiveOption),
+//! [`DefensiveResult`](frame::traits::DefensiveResult)) that can be used to defensively unwrap
+//! values.  This can be used in place of
+//! an `expect`, and again, only if the developer is sure about the unwrap in the first place.
+//!
+//! The primary difference of defensive implementations bring over vanilla ones is the usage of [`debug_assertions`](https://doc.rust-lang.org/reference/conditional-compilation.html#debug_assertions).
+//! `debug_assertions` allows for panics to occur in a testing context, but in
+//! production/release, they will merely log an error (i.e., `log::error`).
+//!
+//! The [`Defensive`](frame::traits::Defensive) trait provides a number of functions, all of which
+//! provide an alternative to 'vanilla' Rust functions:
+//!
+//! - [`defensive_unwrap_or()`](frame::traits::Defensive::defensive_unwrap_or)
+//! - [`defensive_ok_or()`](frame::traits::DefensiveOption::defensive_ok_or)
```

Comment:
In terms of ordering, this para feels to me like it would be better suited swapped with the para above. wdyt? 

```suggestion
//! The [`Defensive`](frame::traits::Defensive) trait provides a number of functions, all of which
//! provide an alternative to 'vanilla' Rust functions, e.g.
//!
//! - [`defensive_unwrap_or()`](frame::traits::Defensive::defensive_unwrap_or)
//! - [`defensive_ok_or()`](frame::traits::DefensiveOption::defensive_ok_or)
```